src/main/java/interpreter/KruskalSnapshot.java

```diff
@@ -0,0 +1,132 @@
+package interpreter;
+
+import lombok.Data;
+import model.Demo;
+import model.Edge;
+import model.Node;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+// Undirected graph algorithm, so edges need matching reverse edge
+// ^ Update: doesn't actually need matching reverse edges but without them it would
+//   make a tree that isn't really connected, it would treat a->b, c->b, c->d as a "tree" for example
+// Ignores demo.end() since this algorithm generates a minimum spanning tree
+// Also ignores demo.start() since Kruskal's algorithm always starts with the smallest edge
+
+@Data
+public class KruskalSnapshot implements Snapshot {
+    /*
+        Kruskal's algorithm:
+            tl;dr Keep adding the lowest cost edge that connects vertices until all vertices are connected
+    */
+    // Each node is mapped to its own tree, i.e. the treeOfNode
+    // Joining two trees sets them to have the same int
+    Map<Node, Integer> nodeToTree = new HashMap<>(); // Each node starts as its own tree, numbered from 0
+    List<List<Node>> treeToNodes = new ArrayList<>(); // Mapping of tree index back to a list of nodes. List index is tree number.
+    List<Edge> chosenEdges = new ArrayList<>(); // What eventually becomes the minimum spanning tree of the graph
+    List<Edge> remainingEdges = new ArrayList<>();
+    int numTrees = 0;
+    boolean canContinue = true;
+
+    public KruskalSnapshot(Demo demo) {
+        demo.getGraph().getNodes().forEach(node -> {
+            this.nodeToTree.put(node, numTrees++);
+            List<Node> tree = new ArrayList<>();
+            tree.add(node);
+            this.treeToNodes.add(tree);
+        });
+        demo.getGraph().getEdges().forEach(edge -> {
+            this.remainingEdges.add(edge);
+        });
+        checkForBidirectionalityOfEdges();
+    }
+
+    private void checkForBidirectionalityOfEdges() {
+        int matches = 0;
```

Comment:
nitpicky but i feel like it would be simpler to just do

```for (Edge edge: remainingEdges) {
   if (!remainingEdges.contains(Edge.of(edge.getEnd(), edge.getStart()))
         throw new InterpreterException...```

Reply:
yes you're right!